### ISSUE TITLE
Collection expression loophole: ability to invoke .Add() method with the parameter type not matching the collection expression element type

### ISSUE THREAD

**Issue by controlflow** [dotnet/csharplang#7724]
**Version Used**: `main` branch

**Steps to Reproduce**:

```c#
using System.Collections;
using System.Collections.Generic;

BadCandidateForCollectionExpression xs = [123]; // no .Add(int)
BadCandidateForCollectionExpression ys = ["abc"]; // not int element type
BadCandidateForCollectionExpression zs = [default]; // ????

class BadCandidateForCollectionExpression : IEnumerable<int>
{
  public void Add(string s) { }

  IEnumerator<int> IEnumerable<int>.GetEnumerator() { yield break; }
  IEnumerator IEnumerable.GetEnumerator() { yield break; }
}
```

**Expected Behavior**:

Impossible to fill the collection expression, no `Add` method can accept the value of `int` type - the element type inferred as a collection expression element type.

**Actual Behavior**:

If the expression is both convertible to inferred element type and can be accepted by some non-`int` `Add` method overload - it compiles successfully.
As the result you have a really weird language context - you can only pass expressions that are implicitly convertible to `int`, but the invoked method accepts `string` value.

I guess this is the poorly written check in frontend, I guess instead of checking the element expression to be convertible to `int` you probably must check the parameter type of invoked `.Add()` method to have type convertible to `int`.

**Comment by Ai-N3rd** [comment]
This should not be supported unless we can find a reasonable use case. Otherwise, It shouldn't be supported for collection expressions.

**Comment by GeirGrusom** [comment]
A behavior I found a bit amusing here is that if you change the parameter into a float, and you try to pass a float to that method it will complain that it will only allow an int, not a float! If you give it an int however, it will silently pass a float instead.

**Comment by Ai-N3rd** [comment]
There should probably be some type of check at compile time. It would simply check if the .Add() method has the same type as the IEnumerable. If not, it cant be used in collections expressions.

**Comment by cston** [comment]
> There should probably be a check at compile time. 

Agreed. Perhaps a diagnostic should be reported if the converted element is not implicitly convertible to the collection iteration type. That should handle the reported case, where there is no `Add` method callable with a value of the iteration type, and it should also handle the case where the preferred `Add` method might be an extension method but another `Add` method is used, such as `zs = [default]` below.

```csharp
BadCandidateForCollectionExpression xs = [123];     // Extensions.Add()
BadCandidateForCollectionExpression ys = ["abc"];   // error: not int element type
BadCandidateForCollectionExpression zs = [default]; // warning: converting 'default' to 'string' for Add(string)

class BadCandidateForCollectionExpression : IEnumerable<int>
{
  public void Add(string s) { }

  IEnumerator<int> IEnumerable<int>.GetEnumerator() { yield break; }
  IEnumerator IEnumerable.GetEnumerator() { yield break; }
}

static class Extensions
{
    public static void Add(this BadCandidateForCollectionExpression c, int i) { ... }
}
```
